Automattic/wp-calypso
# Patterns: Fetch and show list of patterns on `Patterns` page

This pull request is a follow-up of https://github.com/Automattic/wp-calypso/pull/87670 which updates the `Patterns` page in order to retrieve a list of patterns from the API, and renders it server-side. The `/ptk/patterns` API endpoint is only ever called once for each language, even when logged-in, since we cache requests:

![image](https://github.com/Automattic/wp-calypso/assets/594356/1e056aba-4836-4f43-b07b-16b081d610f9)

This pull request also makes the following improvements:

* Adds a new `usePatterns()` hook to retrieve the patterns
* Renames `useDotcomPatterns()` to `useAssemblerPatterns()` to clarify where this hook is used
* Updates the query key of the latter to avoid cache pollution with our new hook
* Updates entry points to use TypeScript
* Adds a document title (which ultimately will have to be different per category)
* Adds [hreflang](https://www.oncrawl.com/oncrawl-seo-thoughts/hreflang-and-seo-5-mistakes-to-avoid/) tags for each supported language
* Adds a [canonical](https://www.oncrawl.com/oncrawl-seo-thoughts/the-rel-canonical-tag-usage-and-pitfalls/) tag

```
<link rel="alternate" hrefLang="ar" href="https://wordpress.com/ar/patterns"/>
<link rel="alternate" hrefLang="de" href="https://wordpress.com/de/patterns"/>
<link rel="alternate" hrefLang="en" href="https://wordpress.com/patterns"/>
<link rel="alternate" hrefLang="es" href="https://wordpress.com/es/patterns"/>
<link rel="alternate" hrefLang="fr" href="https://wordpress.com/fr/patterns"/>
<link rel="alternate" hrefLang="he" href="https://wordpress.com/he/patterns"/>
<link rel="alternate" hrefLang="id" href="https://wordpress.com/id/patterns"/>
<link rel="alternate" hrefLang="it" href="https://wordpress.com/it/patterns"/>
<link rel="alternate" hrefLang="ja" href="https://wordpress.com/ja/patterns"/>
<link rel="alternate" hrefLang="ko" href="https://wordpress.com/ko/patterns"/>
<link rel="alternate" hrefLang="nl" href="https://wordpress.com/nl/patterns"/>
<link rel="alternate" hrefLang="pt-br" href="https://wordpress.com/pt-br/patterns"/>
<link rel="alternate" hrefLang="ru" href="https://wordpress.com/ru/patterns"/>
<link rel="alternate" hrefLang="sv" href="https://wordpress.com/sv/patterns"/>
<link rel="alternate" hrefLang="tr" href="https://wordpress.com/tr/patterns"/>
<link rel="alternate" hrefLang="x-default" href="https://wordpress.com/patterns"/>
<link rel="alternate" hrefLang="zh-cn" href="https://wordpress.com/zh-cn/patterns"/>
<link rel="alternate" hrefLang="zh-tw" href="https://wordpress.com/zh-tw/patterns"/>
<link rel="canonical" href="https://wordpress.com/fr/patterns"/>
```

Related to https://github.com/Automattic/dotcom-forge/issues/5596

#### Testing instructions

1. Run `git checkout add/patterns-list-to-patterns-page` and start your server, or access a [live branch](https://github.com/Automattic/wp-calypso/pull/87726#issuecomment-1957127293)
2. Make sure you are logged out and that Javascript is disabled
3. Assert that patterns are shown in English when loading http://calypso.localhost:3000/patterns
4. Assert that the source code of the page has the right `hreflang` and `canonical` tags
5. Assert that patterns are shown in French when loading http://calypso.localhost:3000/fr/patterns
6. Assert that the title of the page is `WordPress Patterns - WordPress.com` each time
7. Log into a WordPress.com account that is set to English
8. Assert that patterns are shown in English when loading http://calypso.localhost:3000/patterns
9. Assert that patterns are shown in Spanish when loading http://calypso.localhost:3000/es/patterns
10. Assert that you were redirected to http://calypso.localhost:3000/patterns from http://calypso.localhost:3000/es/patterns
